In [2]:
import tensorflow as tf

## Load the mnist dataset

In [3]:
mnist = tf.keras.datasets.mnist

(X_train, Y_train), (X_test, Y_test) = mnist.load_data() #Load data into matrix X and  Y

X_train, X_test = X_train/255.0, X_test/255.0 #Convert the samples from int to float.

11493376/11490434 [==============================] - 6s 0us/step


## Create tf.keras.Sequential model by stacking layers

In [4]:
#Pick an optimizer and loss function for tranining.
model = tf.keras.models.Sequential([ 
    tf.keras.layers.Flatten(input_shape=(28,28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)])

## Create MLP Model

In [5]:
predictions = model(X_train[:1]).numpy() #Model returns a vector of logits/log-odds scores, one for each class.
predictions #View predictions.


If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.



array([[-0.2176262 ,  0.39312714,  0.1899533 ,  0.413311  ,  0.20894721,
        -0.5570416 ,  0.46739113,  0.5999389 ,  0.90800816,  0.08501256]],
      dtype=float32)

In [6]:
tf.nn.softmax(predictions).numpy() #Covert logits to probabilities for each class.

array([[0.05829908, 0.1073763 , 0.0876337 , 0.10956559, 0.08931413,
        0.04151982, 0.11565407, 0.13204613, 0.17968774, 0.07890345]],
      dtype=float32)

In [7]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True) #Return scalar loss for each example.

In [8]:
loss_fn(Y_train[:1], predictions).numpy()

3.1815844

In [9]:
model.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])

In [10]:
model.fit(X_train, Y_train, epochs=5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 2s 39us/sample - loss: 0.3005 - accuracy: 0.9126
Epoch 2/5
60000/60000 [==============================] - 2s 35us/sample - loss: 0.1443 - accuracy: 0.9566
Epoch 3/5
60000/60000 [==============================] - 2s 35us/sample - loss: 0.1076 - accuracy: 0.9675s - loss: 0 - ETA: 0s - loss: 0.1 - ETA: 0s - loss: 0.1073 - accuracy: 0.
Epoch 4/5
60000/60000 [==============================] - 2s 35us/sample - loss: 0.0869 - accuracy: 0.9730
Epoch 5/5
60000/60000 [==============================] - 2s 33us/sample - loss: 0.0748 - accuracy: 0.9764


In [11]:
model.fit(X_train, Y_train, epochs=5) #Adjusts the model params to minimize loss.

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 2s 36us/sample - loss: 0.0654 - accuracy: 0.9795
Epoch 2/5
60000/60000 [==============================] - 2s 36us/sample - loss: 0.0586 - accuracy: 0.9812
Epoch 3/5
60000/60000 [==============================] - 2s 36us/sample - loss: 0.0523 - accuracy: 0.9829
Epoch 4/5
60000/60000 [==============================] - 2s 36us/sample - loss: 0.0494 - accuracy: 0.9833
Epoch 5/5
60000/60000 [==============================] - 2s 35us/sample - loss: 0.0432 - accuracy: 0.9855


## Evaluate Model

In [12]:
model.evaluate(X_test, Y_test, verbose=2) #Check model performance on the test set.

10000/10000 - 0s - loss: 0.0747 - accuracy: 0.9784


[0.07474448034880334, 0.9784]

## Results
Hence, our MLP model is able to provide us with 97.84% accuracy.